python/cudf/cudf/_lib/scalar.pyx

```diff
@@ -60,74 +65,129 @@ cdef class Scalar:
+            if isinstance(value, (np.datetime64, np.timedelta64)):
+                if np.isnat(value) and np.datetime_data(
+                    value.dtype
+                )[0] == 'generic':
+                    raise TypeError("Need a dtype to build a NaT Scalar")
+                else:
+                    dtype = value.dtype
             else:
                 dtype = value.dtype
-
         dtype = np.dtype(dtype)
 
+        # Caching Mechanism
+        self._initialize_cache(value, dtype)
+
+    def _initialize_cache(self, value, dtype):
+        if is_null_host_scalar(value):
+            self._host_value = cudf.NA
+        else:
+            self._host_value = value
+        # need to save the host dtype, because this determines
+        # the codepath for setting the device value later
+        self._host_dtype = dtype
+
+    def _host_value_current(self):
+        return self._host_value is not None
+
+    def _device_value_current(self):
+        return self.uptr._device_uptr != NULL
+
+    def set_device_value(self, value, dtype):
+        valid = not is_null_host_scalar(value)
+
         if pd.api.types.is_string_dtype(dtype):
-            _set_string_from_np_string(self.c_value, value, valid)
+            _set_string_from_np_string(self.uptr._device_uptr, value, valid)
         elif pd.api.types.is_numeric_dtype(dtype):
-            _set_numeric_from_np_scalar(self.c_value, value, dtype, valid)
+            _set_numeric_from_np_scalar(self.uptr._device_uptr,
+                                        value,
+                                        dtype,
+                                        valid)
         elif pd.api.types.is_datetime64_dtype(dtype):
             _set_datetime64_from_np_scalar(
-                self.c_value, value, dtype, valid
+                self.uptr._device_uptr, value, dtype, valid
             )
         elif pd.api.types.is_timedelta64_dtype(dtype):
             _set_timedelta64_from_np_scalar(
-                self.c_value, value, dtype, valid
+                self.uptr._device_uptr, value, dtype, valid
             )
         else:
             raise ValueError(
                 f"Cannot convert value of type "
                 f"{type(value).__name__} to cudf scalar"
             )
 
+    def get_device_value(self):
+        if pd.api.types.is_string_dtype(self.dtype):
+            result = _get_py_string_from_string(self.uptr._device_uptr)
+        elif pd.api.types.is_numeric_dtype(self.dtype):
+            result = _get_np_scalar_from_numeric(self.uptr._device_uptr)
+        elif pd.api.types.is_datetime64_dtype(self.dtype):
+            result = _get_np_scalar_from_timestamp64(self.uptr._device_uptr)
+        elif pd.api.types.is_timedelta64_dtype(self.dtype):
+            result = _get_np_scalar_from_timedelta64(self.uptr._device_uptr)
+        else:
+            raise ValueError(
+                "Could not convert cudf::scalar to a Python value"
+            )
+        return result
```

Comment:
Should this be an internal method as it seems the only use case is using it within the `value` property?

Reply:
Done.